typegraph/typegraph/providers/prisma/type_generator.py

```diff
@@ -124,6 +126,211 @@ def get_where_type(
         else:
             return t.struct(fields).named(name)
 
+    def gen_any_type(self):
+        base_any_type = rename_with_idx(
+            t.union([t.integer(), t.float(), t.boolean(), t.string()]), "base_any_type"
+        )
+        return t.union([base_any_type, t.array(base_any_type)])
+
+    # Example:
+    # t.X() => t.either([t.X(), struct_equals, struct_startsWith, ...])
+    def extend_terminal_nodes_props(self, tpe: t.struct) -> t.struct:
+        any_type = self.gen_any_type()
+
+        # node level
+        term_list = [
+            t.struct({"equals": any_type}),
+            t.struct({"in": any_type}),  # t.array(any_type) ??
+            t.struct({"notIn": any_type}),  # t.array(any_type) ??
+            t.struct({"startsWith": t.string()}),
+            t.struct({"endsWith": t.string()}),
+            t.struct({"contains": t.string()}),
+            t.struct({"gt": any_type}),
+            t.struct({"lt": any_type}),
+            t.struct({"gte": any_type}),
+            t.struct({"lte": any_type}),
+        ]
+
+        # rename with a prefix
+        for i in range(len(term_list)):
+            term = term_list[i]
+            # use the first key
+            prefix = next(iter(term.props))
+            term_list[i] = rename_with_idx(term, prefix)
+
+        node_props = {}
+        for k, v in tpe.props.items():
+            term = t.either(term_list + [undo_optional(v)])
+            not_node = rename_with_idx(t.struct({"not": term}), "not")
+            node_props[k] = t.either([not_node, term]).optional()
+
+        return t.struct(node_props)
+
+    # Examples:
+    # where: { name: { not: { equals: "John" } } }
+    # where: { AND: [ { unique: { gt: 2 } }, { name: { startsWith: "P" }}]}
+    # where: { NOT: { NOT: { name: { startsWith: "P" }} }}
+    # where: { field1: 4, field2: {gt: 3}}
+    def gen_query_where_expr(
+        self, tpe: t.struct, exclude_extra_fields=False
+    ) -> t.struct:
+        tpe = self.get_where_type(tpe)
+        extended_tpe = self.extend_terminal_nodes_props(tpe)
+        extended_tpe = rename_with_idx(extended_tpe, "extended_tpe")
+
+        # define the terminal expression
+        temp_props = {k: v.optional() for k, v in extended_tpe.props.items()}
+        intermediate = t.proxy("???")
```

Comment:
?

Reply:
`???` is a placeholder, can be anything else, I mutate it later to self-loop the expression definition thus allowing infinite nested expressions dynamically.